bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+        if let Some((_, err)) = &self.rack_init_error {
+            return Err(err.clone());
+        }
+        let request_id = match &self.state {
+            State::Uninitialized => return Err(ApiError::NotInitialized),
+            State::Learning { .. } => return Err(ApiError::StillLearning),
+            State::InitialMember { pkg, .. } => {
+                self.request_manager.new_load_rack_secret_req(
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    &self.connected_peers,
+                )
+            }
+            State::Learned { pkg } => {
+                self.request_manager.new_load_rack_secret_req(
+                    now,
+                    pkg.rack_uuid,
+                    pkg.threshold,
+                    &self.connected_peers,
+                )
+            }
+        };
+
+        Ok(request_id)
+    }
+
+    /// Periodic tick to check for request expiration and trigger learner
+    /// attempt peer rotation.
+    ///
+    /// Return any expired request errors mapped to their request id
+    pub fn tick(
+        &mut self,
+        now: Instant,
+    ) -> Result<(), BTreeMap<Uuid, ApiError>> {
+        if let State::Uninitialized = &self.state {
+            return Ok(());
+        }
+        if let Some((request_id, err)) = &self.rack_init_error {
+            return Err(BTreeMap::from([(request_id.clone(), err.clone())]));
+        }
+        let mut errors = BTreeMap::new();
+        for (req_id, req) in self.request_manager.expired(now) {
+            match req {
+                TrackableRequest::InitRack { acks, .. } => {
+                    let unacked_peers = acks
+                        .expected
+                        .difference(&acks.received)
+                        .cloned()
+                        .collect();
+                    let err = ApiError::RackInitTimeout { unacked_peers };
+                    errors.insert(req_id, err.clone());
+                    self.rack_init_error = Some((req_id, err));
+                }
+                TrackableRequest::LoadRackSecret { .. } => {
+                    errors.insert(req_id, ApiError::RackSecretLoadTimeout);
+                }
+                TrackableRequest::LearnReceived { .. } => {
+                    // Nothing to do here, as these are requests from messages
+                    // and not api requests
+                }
+                TrackableRequest::LearnSent { to } => {
+                    if let State::Learning = &self.state {
+                        // Issue the next request
+                        if let Some(next_peer) = self.next_peer(&to) {
+                            let _ = self
+                                .request_manager
+                                .new_learn_sent_req(now, next_peer);
+                        }
+                    }
+                }
+            }
+        }
+        if errors.is_empty() {
```

Comment:
The semantics of this function seem odd.

How about returning a `Result<BTreeSet<Uuid>, ApiError>`:
- `Ok(..)` is a list of requests that have expired (no need to associated them with a `RackSecretLoadTimeout`, because we _know_ that's what they are).  If it's empty, then nothing has expired!
- `Err(..)` is if `self.rack_init_error` is populated

(this also gets cleaner if we do the `enum` strategy from [above](https://github.com/oxidecomputer/omicron/pull/3385/files#r1256445198), because then this function wouldn't know about `self.rack_init_error` and could return `BTreeSet<Uuid>`)

Reply:
I actually find your suggestion more confusing, especially if there can still be a  `self.rack_init_error`, as I am currently operating under the assumption of. The reason is that this bifurcates the errors now. We'd return a set of UUIDs as the `Ok` variant for `RackSecretLoadTimeout` but return an actual `Err(ApiError)` for `RackInitTimeout`.

The caller is going to want to associate it's actual errors to UUIDs so they can be mapped to futures, so I think maintaining this as is will work a bit better.